packages/seacas/libraries/ioss/src/exodus/Ioex_DecompositionData.C

```diff
@@ -2174,6 +1934,249 @@ namespace Ioex {
+                                                    const std::vector<int64_t>& blockSubsetIndex, size_t step,
+                                                    const std::vector<BlockFieldData>& block_data) const
+  {
+    const DecompositionData<INT> *decompData = dynamic_cast<const DecompositionData<INT> *>(m_decompositionDB);
+    Ioss::Utils::check_dynamic_cast(decompData);
+
+    decompData->m_decomposition.show_progress(__func__);
+
+    double *data = reinterpret_cast<double *>(iossData);
+    size_t num_blocks = blockSubsetIndex.size();
+
+    std::vector<int> blockComponentCount(num_blocks);
+    for(size_t blk_seq = 0; blk_seq < blockSubsetIndex.size(); blk_seq++) {
+      blockComponentCount[blk_seq] = block_data[blk_seq].comp_count;
+    }
+
+    std::vector<size_t> fileConnOffset = get_file_offset<INT>(blockSubsetIndex, blockComponentCount);
+    size_t file = m_batchOffset.get_ioss_element_size(blockSubsetIndex);
+
+    decompData->m_decomposition.show_progress("\tex_get_partial_var (elem)");
+
+    if (decompData->m_decomposition.m_method == "LINEAR") {
+
+      for(size_t i = 0; i < blockSubsetIndex.size(); i++) {
+        int64_t blk_seq = blockSubsetIndex[i];
+        const Ioss::BlockDecompositionData& blk = decompData->el_blocks[blk_seq];
+        int64_t id = blk.id();
+        size_t offset = decompData->get_block_element_offset(blk_seq);
+        size_t count  = decompData->get_block_element_count(blk_seq);
+
+        for(size_t comp=0; comp<block_data[i].comp_count; comp++) {
+          size_t var_index = block_data[i].var_index[comp];
+          size_t file_index = fileConnOffset[i] + count*comp;
+
+          int ierr = ex_get_partial_var(filePtr, step, EX_ELEM_BLOCK, var_index, id, offset + 1, count,
+                                        &data[file_index]);
+
+          if (ierr < 0) {
+            Ioex::exodus_error(filePtr, __LINE__, __func__, __FILE__);
+          }
+        }
+      }
+    }
+    else {
+      std::vector<double> file_data(fileConnOffset[num_blocks]);
+
+      for(size_t i = 0; i < blockSubsetIndex.size(); i++) {
+        int64_t blk_seq = blockSubsetIndex[i];
+        const Ioss::BlockDecompositionData& blk = decompData->el_blocks[blk_seq];
+        int64_t id = blk.id();
+        size_t offset = decompData->get_block_element_offset(blk_seq);
+        size_t count  = decompData->get_block_element_count(blk_seq);
+
+        for(size_t comp=0; comp<block_data[i].comp_count; comp++) {
+          size_t var_index = block_data[i].var_index[comp];
+          size_t file_index = fileConnOffset[i] + count*comp;
+
+          int ierr = ex_get_partial_var(filePtr, step, EX_ELEM_BLOCK, var_index, id, offset + 1, count,
+                                        &file_data[file_index]);
+
+          if (ierr < 0) {
+            Ioex::exodus_error(filePtr, __LINE__, __func__, __FILE__);
+          }
+        }
+      }
+
+      decompData->m_decomposition.communicate_entity_data(file_data.data(), data,
```

Comment:
I didn't think we needed to communicate for a "LINEAR" decompositions?  At least for element data...

Both parts of the if block look to be the same...

Reply:
I was following the pattern that had been set up on 307 in Ioss_Decomposition.h for `communicate_element_data()` ..

Reply:
I think you missed the `return` that is in the `LINEAR` part of the if block then...  There is also a copy in that block: (I removed asserts/comments just to shorten):
```
 if (m_method == "LINEAR") {
        auto size = localElementMap.size() * comp_count;
        std::copy(file_data, file_data + size, ioss_data);
        return;
      }
```

Reply:
This function is the batch read version of `get_elem_var()` on line 1319 which does an exodus read from file into a data vector if the decomposition method is LINEAR. Since the generalization is based on subsetted element blocks, it is looping over those individual blocks and reading the data into the right locations just like `get_elem_var()`. I only mentioned the pattern above in `communicate_field_data()` as an example of where you check for the LINEAR decomposition method when doing parallel IO .. the batch equivalent of `communicate_field_data()` is `communicate_entity_data()` on line 653 in Ioss_Decomposition.h which does the early return that you mentioned.

Reply:
OK, I guess I got my brace count off and it looked like both branches of the if closed right before the `communicate_entity_data`.  I can now see that the LINEAR branch does not have the communicate in it.
Sorry for the false alarm.

Reply:
I don't know whether it would be worth pulling all the duplicate code into a function and calling that from each branch of the if.  Looks like only difference is the use of `data` in the linear portion and `file_data` in the other branch...

Reply:
@gsjaardema I have another PR up that refactors and extracts that function